### ISSUE TITLE
Slow page loading/construction with 'large' database

### ISSUE THREAD

**Issue by kormyen** [kormyen/memex#5]
Load divs on scroll?
Delayed div construction?

I believe the slow-ness/lag is due to constructing dom elements rather than dealing with the text data itself (so far...)

**Comment by kormyen** [comment]
Note to self: look at how loading spinner works in https://github.com/kodedninja/orkl

**Comment by joshavanier** [comment]
I'm having this issue. It takes a couple of minutes for the entire thing to load. It's mostly because of lots of DOM reflows and string manipulations. Instead of writing directly to the body, you should try appending everything to a document fragment then append that to the body when everything is ready.

I tried to implement it last night but I messed something up lol. I need to further study your code :)

**Comment by kormyen** [comment]
Wow 'couple of minutes'? That is pretty bad. I'm seeing 6 second loads with 555 entries and thought that was bad enough.

I need to look into this properly as I do not understand DOM related stuff. A quick google says 

"Reflow occurs when you: 
...
resize the window
scroll"
but I once the page (all entry divs constructed) I can freely scroll, resize etc without lag (reflow delay?)

I thought I was doing it right by calculating everything into a string var then applying that all as one update to this.grid.innerHTML... but instead I had a new write to innerHTML for every entry. Just did that now and there is still a delay but it loads in less than a second with 555 entries on my laptop.

I'll push this soon. Completely open to any other improvements or links/resources you recommend.

**Comment by kormyen** [comment]
Pushed

**Comment by kormyen** [comment]
I'll have a look at promises, but wouldn't the append-per-entry-as-ready get me back to the state before the fix above (many edits to innerHTML causing lag?).

Ah, the CORS error (in Chrome?) will be due to the way I am loading the database. It works locally in Firefox however. I did that as a test of just having the database clean text data, rather than including the ```var DATABASE = ``` js. Maybe the Chrome issues make it not worth the hassle.

**Comment by joshavanier** [comment]
There might still be some latency but there would also be perceived improvements in speed since the UI will be rendered right away while the database is still being processed. Having above-the-fold content rendered as soon as possible is good enough to satisfy most users

**Comment by kormyen** [comment]
Ok. Makes sense.
I'll give it a go.

I suppose Memex will need pagination or 'load more' (button or automatic as scrolling) eventually.

**Comment by kormyen** [comment]
I've changed to standard js db, now it'll run fine locally in Chrome (no CORS error).

**Comment by kormyen** [comment]
Haven't got the promise stuff done yet but memex now loads the html/css first (with a preloader anim!) THEN loads/runs the js. Lots better than blank white page for 6 seconds from a couple of days ago.

**Comment by kormyen** [comment]
Been trying to get it so that entries are displayed as they are computed but the page stays unresponsive (due to reflows I think) until (DOM?) additions/modifications are complete defeating the purpose.

On top of this it takes much longer than just computing all new html and adding/replacing it once.

Open to advice but as my current understanding is going it is in the best state it can be atm:

- instantly show bg and animated loading anim while db parses and html is computed, then added to page
- loading anim removed, results displayed.

Current failed messy testing state here https://github.com/kormyen/memex/commit/1bb09fe77fa7a2fc8bbd6bbd33bcc25b238878ab

**Comment by kormyen** [comment]
Continuing with promise setup https://github.com/kormyen/memex/commits/promise

**Comment by kormyen** [comment]
Ended up removing promises as they didn't change anything other than code layout (no change to speed or responsiveness).

In the process thought I made a nice simple benchmark-er (seer.js), refactored everything and learned about promises.

The bottleneck now is actually the masonry layout call (20 to 90% of the delay is there, longer for more articles). I could try to write a minimal/faster replacement but not a high priority for me atm. Similarly it would be nice to make my own type icons as font-awesome is comparatively large and takes a moment to load.

![benchmark](https://user-images.githubusercontent.com/836507/48110951-032be880-e2b4-11e8-8599-367174ce6639.JPG)

Calling this issue complete.

- load times are faster (previously 500~ articles in 6~ seconds, now 1~ second)
- initial load displays background and preload anim immediately (instead of white page while loading)
- for queries that are not 'quick' preload anim displays as immediate response (instead of no change) while loading
